assembly/src/compiled_library.rs

```diff
@@ -0,0 +1,167 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+use vm_core::{
+    crypto::hash::RpoDigest,
+    mast::{MastForest, MerkleTreeNode},
+};
+
+use crate::{
+    ast::{FullyQualifiedProcedureName, ProcedureIndex, ProcedureName, ResolvedProcedure},
+    LibraryPath, Version,
+};
+
+// TODOP: Refactor `FullyQualifiedProcedureName` instead, and use `Span<FQDN>` where needed?
+pub struct CompiledFullyQualifiedProcedureName {
+    /// The module path for this procedure.
+    pub module_path: LibraryPath,
+    /// The name of the procedure.
+    pub name: ProcedureName,
+}
+
+impl CompiledFullyQualifiedProcedureName {
+    pub fn new(module_path: LibraryPath, name: ProcedureName) -> Self {
+        Self { module_path, name }
+    }
+}
+
+impl From<FullyQualifiedProcedureName> for CompiledFullyQualifiedProcedureName {
+    fn from(fqdn: FullyQualifiedProcedureName) -> Self {
+        Self {
+            module_path: fqdn.module,
+            name: fqdn.name,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CompiledProcedure {
+    pub name: ProcedureName,
```

Comment:
A larger issue with the current state of the PR is that most (all?) names are assumed to be spanned. For example, a `CompiledProcedure`'s name is a `ProcedureName`, which wraps an `Ident`, which stores a `Span`. A `FullyQualifiedProcedureName` also has a `span`; even `components` of a `LibraryPath`. The issue as I understand it is that once compiled, spans should be erased since they refer to a different file from whichever module will depend on that library, and we don't have *e.g.* a `FullyQualifiedSpan` which would encode which file the `Span` comes from.

The "all names have a span" assumption is everywhere in the assembler, and so fixing it is a significant change, and would likely touch just about every part of the assembler (unless I'm missing something). It is even challenging to remove spans from names of new "compiled" types (such as `CompiledProcedure`), because a `CompiledProcedure` is manipulated in the same code paths as `Procedure`s, whose errors expect spans.

Reply:
I was thinking that for things like `FullyQualifiedProcedureName` we would just set the `span` to `SourceSpan::default()` when span info is not available. In fact, I think we already have constructors for many such structs which don't require span info (e.g., `ProcedureName::new()` just takes a string reference).

Or would doing this cause some kind of a problem?